### ISSUE TITLE
panic in Writer run loop

### ISSUE THREAD

**Issue by Masterlvng** [segmentio/kafka-go#517]
**Describe the bug**

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x9e7650]

goroutine 85 [running]:
github.com/segmentio/kafka-go.(*Writer).run(0xc0000f8b40)
        /data/orange-ci/workspace/vendor/github.com/segmentio/kafka-go/writer.go:488 +0x560
created by github.com/segmentio/kafka-go.NewWriter
        /data/orange-ci/workspace/vendor/github.com/segmentio/kafka-go/writer.go:272 +0x2bd
```

**Kafka Version**
0.10

kafka-go version: v0.3.4
```go
if len(partitions) != 0 {
    selectedPartition := w.config.Balancer.Balance(wm.msg, partitions...)
    writers[selectedPartition].messages() <- wm
} 
```
it seams that the selectedPartition is not in writers, but i dont know why.

**Comment by dominicbarnes** [comment]
@Masterlvng do you see this error in the latest stable version? (v0.3.10)

It looks like this might happen after a rebalance, but it doesn't _seem_ like it should be possible to panic here in the latest code. I have started a branch which would address the panic, but if the newest version already prevents this from failing in the first place we can address your problem.

**Comment by Masterlvng** [comment]
Why do we need to check partition again?

**Comment by Masterlvng** [comment]
Could you please tell me why is might happen after a rebalance? Actually, the kafka rebalanced at that time.

**Comment by dominicbarnes** [comment]
The `writers` map is updated after each rebalance, which happens in [this code](https://github.com/segmentio/kafka-go/blob/master/writer.go#L456-L475). The panic indicates that the selected partition for that message doesn't appear to be in the same list that the broker communicated back to us [here](https://github.com/segmentio/kafka-go/blob/master/writer.go#L505-L529), but that doesn't _seem_ like it should happen as we are choosing `selectedPartition` from the same list as `partitions` which should be consistent in this same iteration of that `for` loop.

Do you have any custom configuration for the `Balancer`? The default `RoundRobin` doesn't seem like it would return a partition outside the list provided, but it's definitely worth checking out.

**Comment by Masterlvng** [comment]
I used LeastBytes balancer. It my select wrong partition when partition changed
for example
```go
	b := &LeastBytes{}
	b.Balance(Message{Key: []byte("key")}, []int{1, 2, 3}...)
	b.Balance(Message{Key: []byte("key")}, []int{1, 2, 3}...)
	part := b.Balance(Message{Key: []byte("key")}, []int{1, 2}...)
	if part == 3 {
		t.Errorf("unexpect appear")
	}
```

**Comment by dominicbarnes** [comment]
Great find! I'll take a look at your PR sometime today.

**Comment by dominicbarnes** [comment]
I've merged #523 which at least avoids the panic with a proper error message. As such, I'll close this issue, but will follow up with #519 in order to fix the root cause of the panic.
